### ISSUE TITLE
Add RegularJumps

### ISSUE THREAD

**Issue by Gaussia** [JuliaDiffEq/DiffEqBiological.jl#42]
Right now we create a set of `ConstantRateJump`s, possibly including a few `VariableRateJump`s.

However recently `RegularRateJumps` have been added. We could add this to the `reaction_network` DSL as well. Looking at the syntax the difference is that for a `RegularRateJumps` you put everything into only one function each for rate and affect, it should be no problem doing the same in `reaction_network`. 

We would have to decide which type of `JumpProblem` should be generated by default though.

**Comment by isaacsas** [comment]
@Gaussia In addition to `RegularJumps`, would you think about adding support for the new `MassActionJump` type? Anyone running an SSA would see substantial performance gains if any mass action reactions were encoded through that jump type. Non-mass action reactions (Hill eqs and such) can still be included as constant jumps. 

I'm happy to help if you have questions, and/or provide a MWE on how to setup a hybrid system if this is something that would be of interest to you.

**Comment by Gaussia** [comment]
`MassActionJump` is a new addition? I looked in the Jump part of DiffEq and it seemed like that, but found nothing in the docs.

Definitely sounds like something we should add in. Is there any resource on how these actually look right now. If you tell me what you need to run efficient SSAs (something I would be happy to use), I can ensure that that is stored somewhere in the reaction_network type.

**Comment by isaacsas** [comment]
Yup it is new. Nothing in the docs yet but I just started working on that, so hopefully soon!

Basically there is a new `MassActionJump` type that encodes any jumps that can be represented through a reactant stoichiometry and net stoichiometry. I think an example is easier to understand, so here is a gist with one: [here](https://gist.github.com/isaacsas/d5129e06cfb32fdf7c700321f69b66e6) The script implements the system both with a `reaction_network` and with a `MassActionJump`.

I think the newly added `reactions` field to the generated network should have all the info needed to create a `MassActionJump`. If you have non mass action reactions too (like a Hill equation) they can still be passed as a vector/tuple of additional `ConstantJumps` (just like before). So basically this just lets you tag mass action reactions, and then handles them internally more efficiently. 

PS - One gotcha; I'm not sure how DiffEqBiological currently handles scaling reaction rates for dimer or trimer type reactions. By default `MassActionJumps` assumes rates are as described in Gillespie's 1977 paper (i.e. a rate is the probability per time that one individual set of the reactants comes together and reacts). So for a reaction like 3A -> something at rate k, the rate law it uses is `k*A*(A-1)*(A-2)/6`. If you want it to assume the rate is `k*A*(A-1)*(A-2)` you can pass in the optional kwrag, `scale_rates=false` to the constructor.

**Comment by isaacsas** [comment]
Just to update; the DiffEq docs have been updated to discuss `MassActionJump`s. In particular, the tutorial now shows how to use a `MassActionJump` for the SIR model, and how to couple it with `ConstantRateJumps` for non-mass action jumps.

**Comment by Gaussia** [comment]
Reaction rates are currently handled as k*A*(A-1)*(A-2)/6 for SSA and k*A^3/6 for ODE/SDEs.

Right now or (tuple?) of Jumps is a mix of Constant and Variable Rate Jumps. Would it be mest to change this to a mix of Mass Action, Constant and variable RateJumps (assigning each reaction to the strictest category), or might we risk losing any generality in this case. If that is the case should a second set with this variation?

**Comment by ChrisRackauckas** [comment]
MassActionJumps is an optimization. Any ConstantRateJump can be made a MassActionJump to make it faster. Also, it handles the `kA(A-1)(A-2)/6` inside.

**Comment by isaacsas** [comment]
Just to clarify; any `ConstantRateJump` that can be represented as a normal mass action reaction can be represented as a `MassActionJump`. Things like Hill functions would still need to be `ConstantRateJumps`.


Gaussia, since you already handle the scaling you can just pass the `MassActionJump` constructor the optional `scale_rates=false` flag. That will turn off the scaling so it isn't applied twice.

**Comment by ChrisRackauckas** [comment]
The scaling is only handled during the construction of a `ConstantRateJump`, so if a `MassActionJump` is used then `scale_rates` shouldn't be touched.

**Comment by isaacsas** [comment]
Got it, then everything should work by default.

**Comment by Gaussia** [comment]
I plan to add this as soon as I can get Atom working again. I will ensure that the jumps in the `jumps::Tuple{AbstractJump,Vararg{AbstractJump}}` field that can be represented with a MassActionJump is such a mass action jump (and that the rate is not dependant on things like /2!).

A question however: Is it possible for the rate in a `MassActionJump`
- To be non constant. Can it e.g. be a hill function of some other variables?
- To be time dependant (like a VariableRateJump)?

Also, it seems like `MassActionJump` is, like `RegularJump` and unlike `ConstantRateJump` and `VariableRateJump` a set of several jumps rather than a single Jump (is there some reason for this inconsistency?). Also, how does `RegularJump` fit into all of this and can we use it somehow? 
Are the final goal to have a tuple with several `ConstantRateJump`s and `VariableRateJump`s, as well as a single `RegularJump` and a single `MassActionJump`, each containing several jumps?
(a closer look at the code tells me that there exists a so called `JumpSet`, should we use that instead of a tuple of `AbstractJump`s?)

Finally it seems like the `RegularJump` and `MassActionJump`s are yet not part of the package, since I cannot get them even if I update. But maybe there is some Git-way around this?

**Comment by ChrisRackauckas** [comment]
>Finally it seems like the RegularJump and MassActionJumps are yet not part of the package, since I cannot get them even if I update. But maybe there is some Git-way around this?

Oh! I needed to setup a release. Thanks for reminding me.

>Also, it seems like MassActionJump is, like RegularJump and unlike ConstantRateJump and VariableRateJump a set of several jumps rather than a single Jump (is there some reason for this inconsistency?). Also, how does RegularJump fit into all of this and can we use it somehow?
Are the final goal to have a tuple with several ConstantRateJumps and VariableRateJumps, as well as a single RegularJump and a single MassActionJump, each containing several jumps?

MassActionJumps are just optimizations for ConstantRateJumps, so they have to not be time dependent. They have to use mass action, so they cannot include things like hill functions. It is not a RegularJump because a RegularJump allows for arbitrary rate functions. But a RegularJump doesn't allow for arbitrary effects: the effects have to be changes to the values of the state vector `u` (where as the other jumps can do literally anything like change the ODE solver tolerance). I think it could be good to have a developer triage call so that way we can discuss the implementation details on them. 

But the key difference is that a RegularJump is not time-adapted. All of the three other jumps are for specifying time-adapted jumps, i.e. jump times are calculated and jumps are performed exactly. RegularJumps are required for things like tau-leaping. What we should build DiffEqBiological towards is hybrid algorithms. The way this can be done is by setting up ConstantRate/MassAction/VariableRate jumps for each reaction, and also create a row of the RegularJump for each reaction. Then we have a boolean array that says whether in this step it should be treated as discrete or continuous, and then multiply by that boolean (or !boolean) to get the rates. If we then make that a one-hot vector we can also make it choose between jump/tau-leaping/SDE, or jump/tau-leaping/ODE (no reason to mix the two)

They are all slightly different. I don't like that we have expanded the field, it is more complicated, but they do all serve a distinct purpose and the optimizations provided by the MassActionJump are not small. 

>(a closer look at the code tells me that there exists a so called JumpSet, should we use that instead of a tuple of AbstractJumps?)

Yes.

**Comment by Gaussia** [comment]
I think I start to see a little clearer, correct me if I am wrong:
- `MassActionJump` is a _very_ strict type of jump. Only allows for a reaction on the form `Sum(c_i*S_i) --> Sum(d_i,P_i)`, where the reaction rate is constant. Also calculates the actual mass action rates for the user as an extra service (especially nice if any(c_i,d_i > 1)).
- `ConstantRateJump` A very general type of jump which does not need to change the values of u, but might e.g. change parameter values. However the rate is not time dependant.
- `VariableRateJump` Like `ConstantRateJump` but the rate may be time dependant.
- `RegularRateJump` Like the constant rate jump but the actual effect of the jump has to make changes to u only. Also this is not time-adapted which all makes it run on a slightly different set of algorithms.

Looking at the `reaction_network` type, what should be done first is to
- Ensure that the current `jump` vector produced contains `MassActionJump`s instead of `ConstantRateJump`s for those reaction were this is possible.
- In addition add a field `regular_jumps`, which is an Vector containing a `RegularJump` for each reaction.

**Comment by ChrisRackauckas** [comment]
Yup that's a good summary. We might want to add that bullet point form here:

http://docs.juliadiffeq.org/latest/types/jump_types.html

I would just expand that `VariableRateJump` requires an ODE solver and is much less efficient, so there's a cost to its generality.

**Comment by Gaussia** [comment]
There is always a market for good documentation...

Using `VariableRateJump` is the only way to make jump which rate depends on time, right? I think it is a relatively small category of models which requires this, in which case one will have to bite that bullet. But it is probably good to ensure that is known so people can avoid time dependant rates when possible.

**Comment by isaacsas** [comment]
For stochastic models, there are SSAs extensions that try to efficiently allow time-dependent rates (and delays). I'm not super knowledgable about them, but ultimately we could look into them. They might still require a jump type with more structure than `VariableRateJump`. It might also be that they would need a new solver (like `SSAStepper`) to actually run efficiently, but that could be an opportunity to think about how to get more efficient timesteppers that allow hybrid systems.

I think having time-dependent rates is actually quite useful. If one wants to model anything within a growing cell it would be reasonable to assume the volume is changing. This would lead to time-dependent zero and second order rate constants (through the changing volume).

**Comment by ChrisRackauckas** [comment]
>For stochastic models, there are SSAs extensions that try to efficiently allow time-dependent rates (and delays). I'm not super knowledgable about them, but ultimately we could look into them. They might still require a jump type with more structure than VariableRateJump. It might also be that they would need a new solver (like SSAStepper) to actually run efficiently, but that could be an opportunity to think about how to get more efficient timesteppers that allow hybrid systems.

They have to be equivalent to the `VariableRateJump` in some way because the integral equation needs to be solved. In most interesting cases the user won't know analytical solution to the integral, in which case it needs to be numerically solved simultaneously and then event handling of some form gets involved. We could add an optimization so that way if it's simple time-dependent then the user could give the integral solution as a function of `(u,p,t)`. But I haven't gone that route because my main goal was to allow continuous variables in `u` be part of the rates.

But that's for quasi-exact jumping with time-dependent rates. I think in most cases with time-dependent rates you'd be willing to give up exactness and use something like tau-leaping, in which case time-dependence is perfectly fine.

**Comment by isaacsas** [comment]
> In most interesting cases the user won't know analytical solution to the integral, in which case it needs to be numerically solved simultaneously and then event handling of some form gets involved. 

Aren't there also thinning approaches that get used (where you define an extra "fictitious" jump process so that the overall jump rate is constant)? When the "fictitious" jump gets selected you do advance the time but don't actually change the state at all. 

> But that's for quasi-exact jumping with time-dependent rates. I think in most cases with time-dependent rates you'd be willing to give up exactness and use something like tau-leaping, in which case time-dependence is perfectly fine.

That could be true; it may be that the time-dependent rate SSAs are sufficiently slow that one is better off using tau-leaping with a small timestep / tolerance.

**Comment by ChrisRackauckas** [comment]
>Aren't there also thinning approaches that get used (where you define an extra "fictitious" jump process so that the overall jump rate is constant)? When the "fictitious" jump gets selected you do advance the time but don't actually change the state at all.

There is a fictitious jump method we could implement for these, yes. Let's get an issue open for that.

**Comment by isaacsas** [comment]
@Gaussia To go along with supporting SSAs we'll eventually need to be able to build dependency graphs from the reaction_network structure. I was looking at this example you gave yesterday:

```
network = @reaction_network rnType  begin
           0.01, (X,Y,Z) --> 0
           hill(X,3.,100.,-4), 0 --> Y
           hill(Y,3.,100.,-4), 0 --> Z
           hill(Z,4.5,100.,-4), 0 --> X
           hill(X,2.,100.,6), 0 --> R
           hill(Y,15.,100.,4)*0.002, R --> 0
           20, 0 --> S
           R*0.005, S --> SP
           0.01, SP + SP --> SP2
           0.05, SP2 --> 0
       end;
```
When I look at:

```network.reactions[4].substrates```

I get 

``` 0-element Array{DiffEqBiological.ReactantStruct,1} ```

Is there any easy way to tell what species are needed for the rate law? I would have thought this is encoded by `substrates`, but I guess that would require writing this reaction as `X --> X + Y` using the Hill function rate law?

Also, is it possible to add a field like `has_mass_action_rate_law` so it is easy to tell if the underlying rate law can fit into a `MassActionJump` or needs a constant jump?

**Comment by isaacsas** [comment]
> There is a fictitious jump method we could implement for these, yes. Let's get an issue open for that.

I opened one in DiffEqJump (though I don't think this is something that I'll have time to work on in the near future).

**Comment by ChrisRackauckas** [comment]
>though I don't think this is something that I'll have time to work on in the near future

Yeah, that's the issue. We have something that works, so if you need it it's already there 👍 . And I linked to a paper that showed that with MATLAB's event handling they could improve timings by 3x (and have it be more exact, but IMO that's an issue with MATLAB's tolerances being so high in order to get decent timings). So in the end, these other methods are a small timing improvement for something that's not all that common... so I'll get to it eventually...

**Comment by Gaussia** [comment]
Lets first consider the fourth reaction:
```julia
hill(Z,4.5,100.,-4), 0 --> X
```
The substrate array is empty because there are no substrates in the reaction (at least as I would like to consider the reaction, the substrates are the stuff on the left of the arrow and also the stuff that is consumed by the reaction).

If you are interested in what reactants the reaction depends on you should look at the rate field. In this case it would be an expression like:
```julia
4.5*100/(100^4+Z^4)
```
By parsing this we can check which species it contains (we probably should make functions for this, and maybe making it a separate field in the reaction type, containing an array of symbols corresponding to the species the reaction depends on).

Further, say we have this reaction:

```julia
mm(Z,2.0,5), X+Y -->XY
```
both `X` and `Y` would be mentioned in among the substrates, but not `Z`. The reaction rate field would now look like 
```julia
X*Y*2.0*Z/(5+Z)
```
however, this is not it. We have a special arrow if one do not want to mass action for a specific reaction:
```julia
mm(Z,2.0,5), X+Y ⟾XY
```
this would yield a rate 
```julia
2.0*Z/(5+Z)
```
However `X` and `Y` would still be among the substrates since they are the stuff being consumed by the reaction.

The ⟾ is probably used very rarely (I think I have used it once), but partly due to this one need to keep in mind not to look at the substrates but the rate when looking at dependency.

There actually already exists a field `use_mass_kin` in reactions, but these are tied to whenever the ⟾ arrow is used and the reaction rates should be corrected for mass kinetics or not (so this field have a different purpose). Such a field could be added I guess? However, whenever a `MassActionJump` or a `ConstantRateJump` should be used should be done in the reaction_network?

**Comment by ChrisRackauckas** [comment]
>There actually already exists a field use_mass_kin in reactions, but these are tied to whenever the ⟾ arrow is used and the reaction rates should be corrected for mass kinetics or not (so this field have a different purpose). Such a field could be added I guess? However, whenever a MassActionJump or a ConstantRateJump should be used should be done in the reaction_network?

That's still necessary since you're still constructing the `ConstantRateJump` in those cases. Basically, in the case where you have `p, X + Y -> Z` you create a new stoichiometry in the `MassActionJump` instead of a new `ConstantRateJump`. Those should be prevalent enough for the optimization to truly catch most of a large network in many cases (if not, the new SSAs will!)

**Comment by isaacsas** [comment]
One last question; is it true that the net stoichiometry (i.e. all species that change value from the reaction occurring) can be determined from just `products` and `substrates`? In that case I think the two new fields I mentioned, `has_mass_action_rate_law` and `rate_law_species` would be all I need to build dependency graphs for the new SSAs.

**Comment by Gaussia** [comment]
In this case `use_mass_kin` is not so much an indicator of whenever the reaction is a pure mass kinetics reaction or not, as a way to tell the program that it should multiply the designated rate by the concentrations of the substrates (which is almost always the case), taking higher order terms in account if present.

Yes, net stoichiometry can be determined from that (in fact their exists a function, `function get_stoch_diff(reaction::ReactionStruct, reactant::Symbol)` which calculates this for a reactant in a reaction. This is used to determine what the updates in the jumps generated by the equation should be).

**Comment by Gaussia** [comment]
Also, is it a quick/slow thing to generate this dependency graph? Maybe it should be calculated and added to the `reaction_network` when it is created?

**Comment by ChrisRackauckas** [comment]
>Also, is it a quick/slow thing to generate this dependency graph? Maybe it should be calculated and added to the reaction_network when it is created?

I can image it taking quite a bit of memory in some cases. For these kinds of things what I've been doing with ModelingToolkit.jl is making users query for the extra computations. Maybe adding a convenience function that queries for everything could be a good idea for most cases, but it's not hard to find examples where you don't want that.

**Comment by ChrisRackauckas** [comment]
Let's keep dependency graph discussions in https://github.com/JuliaDiffEq/DiffEqBiological.jl/issues/43

**Comment by Gaussia** [comment]
How should the constructor look like, right now I have
```julia
rj = RegularJump(rate,c,dc;constant_c=true)
```
just checking whenever this is correct, with `...;constant_c=true`. 
@isaacsas @ChrisRackauckas

**Comment by isaacsas** [comment]
I haven't used `RegularJump`s yet myself, so I'm not sure of the interface details.

**Comment by ChrisRackauckas** [comment]
Yup, that is correct. Biological networks all have a constant `c` matrix (the stoichiometry, i.e. what changes when there's a jump). This enables an optimization where it's just stored and multiplied.
